bookFinder/script.js

```diff
@@ -0,0 +1,25 @@
+const captureSearchValue = () => {
+    const searchValue = document.getElementById('search-bar').value;
+    return searchValue;
+};
+
+const filterBooks = (searchString, library) => {
+    const listOfBooks = flattenObjectValuesIntoArray(library);
+    const matchingBooksIndices = listOfBooks
+        .map((book, index) => {
+            if (book.includes(searchString)) {
+                return index;
+            }
+        })
+        .filter((element) => element !== undefined);
+    return matchingBooksIndices.map((item) => library[item]);
+};
```

Comment:
So this code is:
- mapping over the array to convert to index if it's "in"
- then filtering based on that determination

Why not just do it together?

If you're finding that you want to do a map & filter together, you're probably gonna want a `reduce`

```suggestion
const filterBooks = (searchString, library) => {
    const listOfBooks = flattenObjectValuesIntoArray(library);
    const matchingBooks = listOfBooks
        .reduce((matchedBooks, book, index) => {
            if (book.includes(searchString)) {
                matchedBooks.push(library[index]);
            }
            
            return matchedBooks;
        }, [])
   return matchingBooks;
};
```

Reply:
Oh my god that's genius! I have only used `reduce` with numbers, but I see what you did there! You used the accumulation parameter as an array and initialized it as such in the `reduce` method's 2nd parameter. 🤯 #mindBlown

Reply:
Building on this, the way to think about reduce is "I want to transform an array into a new object that requires iterating over the array"

I think Matt's code is super clean here, and is a great learning moment around reduce, but I also think your original code is still pretty good. 

In C# there's an array method called "compact" that basically does what your filter was doing, and mapping things to new values or undefined/null and then filtering out the bad values (ie compacting the array) is a perfectly viable strategy and pattern and you'll likely see it happen a lot. 

Most newer developers aren't comfortable with reduce and find it hard to read so they'll lean heavier on the other methods. It's my hope that you'll skip that part of development quickly and get comfy with reduce early. 

Your quality of code is improving very rapidly by the way. Nice work